### PR TITLE
test: fixed initial `[WILDCARD]` not matching empty the on first line

### LINKED ISSUE #5324: "run_v8_help" integration test fail - "aarm64"
`run_v8_help` integration test failed on `aarm64` builder.
To my knowledge this test has been running fine on my `x86_64` systems.

It could be because it was using the normal release copy of the lib `librustyv8.a` file rather than the debug copy...

Will try reproduce outside of build on a rpi4 soon & provide a trace, test some fixes, etc.

Log looks like it should match the regex, It has `Synopsis:` followed by `d8` later, maybe something funky with spaces, symbols, or newlines only on aarch64 systems...

Error log snippet:
```
[cli/tests/integration_tests.rs:2773] part = "\nSynopsis:\n"
[cli/tests/integration_tests.rs:2773] i = 1
[cli/tests/integration_tests.rs:2782] "exit false " = "exit false "
[cli/tests/integration_tests.rs:2782] i = 1
OUTPUT
Synopsis:
  shell [options] [--shell] [<file>...]
  d8 [options] [-e <string>] [--shell] [[--module] <file>...]
  -e        execute a string in V8
...
OUTPUT
EXPECTED
[WILDCARD]
Synopsis:
[WILDCARD]d8[WILDCARD]
EXPECTED
thread 'run_v8_help' panicked at 'pattern match failed', cli/tests/integration_tests.rs:2740:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

[Full log link](https://logs.nix.ci/?key=nixos/nixpkgs.87653&attempt_id=7e42df57-c7ab-4e0b-be71-a3ea6bb33b3c)

### PR DESCRIPTION
**Pull request by 06kellyjac**
Resolves #5324

This isn't a particularly ideal fix so I strongly welcome recommendations/suggestions

---

`aarch64`'s output from the following command is different to `x86_64`:
`deno repl --v8-flags=--help`

`x86_64` output starts with:
```
SSE3=1 SSSE3=1 SSE4_1=1 SSE4_2=1 SAHF=1 AVX=1 FMA3=1 BMI1=1 BMI2=1 LZCNT=1 POPCNT=1 ATOM=0
Synopsis:
  shell [options] [--shell] [<file>...]
  d8 [options] [-e <string>] [--shell] [[--module] <file>...]
...
```

`aarm64` output starts with:
```
Synopsis:
  shell [options] [--shell] [<file>...]
  d8 [options] [-e <string>] [--shell] [[--module] <file>...]
...
```

match output:
```
[WILDCARD]
Synopsis:
[WILDCARD]d8[WILDCARD]
```

With the current logic the first wildcard can't match to nothing & continue it's checks
This fix allows the first line wildcard to match nothing

### CODE REVIEW

File: cli/tests/integration_tests.rs [hunk 1]
```diff
@@ -2764,6 +2764,14 @@ mod util {
       return false;
     }
 
+    let pattern_first_line = pattern.split("\n").collect::<Vec<&str>>()[0];
+    // If the first line of the pattern is just a wildcard the newline character
+    // needs to be pre-pended so it can safely match anything or nothing and
+    // continue matching.
+    if pattern_first_line == wildcard {
```
Comment (ry) [inline cli/tests/integration_tests.rs:2771]: maybe use `pattern.lines().next() == Some(wildcard)`

Comment (06kellyjac) [reply cli/tests/integration_tests.rs]: Yeah that looks a lot cleaner

**Comment by 06kellyjac** [comment]
It was easy to get arm building, just trying to get as many of the tests as possible running in the nixpkgs check phase :sweat_smile: 

https://github.com/NixOS/nixpkgs/pull/87653

After this is merged into a release the only other holdup issues for running all of the tests are:

* `benchmark_test`
  * `strace` AFAIK can't be ran in the build environments so that'll always skip
* `std_tests`
  * Runs a single deno test that has several benchmark functions
  * `forAwaitFetchDenolandX10` & `promiseAllFetchDenolandX10is` fail because during the check phase there's no internet access, maybe use a local server or split the tests so it's easier to ignore them?
* `_017_import_redirect` `https_import` `js_unit_tests` `lock_write_fetch`
  * All try to write to `target/debug` which is read-only protected at this point
* `upgrade_in_tmpdir` & `upgrade_with_version_in_tmpdir`
  * Will still fail on any nixos machine (`x86_64` or `aarch64`) because the binary needs ELF patching
  * Not a big deal as self updating binaries will never work on nix/nixos bc the `deno` bin will be read-only.
  * These will always be skipped or [patched out](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/networking/cluster/terraform/default.nix#L123-L129)
  * Are there plans for a `rustup` style `deno` version manager? Nix helps me manage interpreters on my machine, but a massive nightmare with node & python is managing versions

File: cli/tests/integration_tests.rs [hunk 2]
```diff
@@ -2764,6 +2764,13 @@ mod util {
       return false;
     }
 
+    // If the first line of the pattern is just a wildcard the newline character
+    // needs to be pre-pended so it can safely match anything or nothing and
+    // continue matching.
+    if pattern.lines().next() == Some(wildcard) {
+      s.insert_str(0, "\n");
+    }
+
```
Comment (bartlomieju) [inline cli/tests/integration_tests.rs:2773]: Can this bit be exercised in test (`test_pattern_match`)?

Comment (06kellyjac) [reply cli/tests/integration_tests.rs:2773]: I'll get on that

Comment (06kellyjac) [reply cli/tests/integration_tests.rs:2773]: Tests added

**Comment by 06kellyjac** [comment]
I'll rebase it now

**Comment by 06kellyjac** [comment]
As a note I slightly take back saying "It was easy to get arm building" :sweat_smile: I tried building deno on aarch64 without nix. I think the big kicker is aarch64 kernel raspbian still has almost completely 32bit userland :/
I've got a full aarch64 kernel & userland on an Arch partition now so I'll have to give it a crack again.

**Comment by 06kellyjac** [comment]
The previous commit failed due to `lock_check_ok` having extra lines for some reason - and only on linux not darwin
linux - https://github.com/denoland/deno/runs/694305340
darwin - https://github.com/denoland/deno/runs/694305304

```
target_dir /home/runner/work/deno/deno/target/release
root path /home/runner/work/deno/deno/cli/..
deno_exe path /home/runner/work/deno/deno/target/release/deno
tools/http_server.py starting...
target_dir /home/runner/work/deno/deno/target/release
output path /home/runner/work/deno/deno/cli/../cli/tests/003_relative_import.ts.out
OUTPUT
Download http://127.0.0.1:4545/cli/tests/003_relative_import.ts
Download http://127.0.0.1:4545/cli/tests/subdir/print_hello.ts
Compile http://127.0.0.1:4545/cli/tests/003_relative_import.ts
Hello

OUTPUT
EXPECTED
Hello

EXPECTED
```

---

Could well be a bug?

**Comment by 06kellyjac** [comment]
Ok windows outputted the original output without the `Download... Compile...` lines

https://github.com/denoland/deno/pull/5420/checks?check_run_id=710548792

I'm going to move it to a `[WILDCARD]` above the output.

Is this intended behavior?

**Comment by bartlomieju** [comment]
> Ok windows outputted the original output without the `Download... Compile...` lines
> 
> https://github.com/denoland/deno/pull/5420/checks?check_run_id=710548792
> 
> I'm going to move it to a `[WILDCARD]` above the output.
> 
> Is this intended behavior?

Yes, integration tests were changed recently to capture both stdout and stderr

**Comment by bartlomieju** [comment]
> It was easy to get arm building, `rusty_v8` had a static library ready, and cargo & rust make things easy.
> Just trying to get as many of the tests as possible running in the nixpkgs check phase 😅
> 
> [NixOS/nixpkgs#87653](https://github.com/NixOS/nixpkgs/pull/87653)
> 
> After this is merged into a release the only other holdup issues for running all of the tests are:
> 
> * `benchmark_test`
>   
>   * `strace` AFAIK can't be ran in the build environments so that'll always skip
> * `std_tests`
> 

Benchmarks are only run on Ubuntu so it shouldn't be a problem
  
>   * Runs a single deno test that has several benchmark functions
>   * `forAwaitFetchDenolandX10` & `promiseAllFetchDenolandX10is` fail because during the check phase there's no internet access, maybe use a local server or split the tests so it's easier to ignore them?

Do they download from `deno.land`? If so we should change that to use local HTTP server (`tools/http_server.py`)

> * `_017_import_redirect` `https_import` `js_unit_tests` `lock_write_fetch`
>   
>   * All try to write to `target/debug` which is read-only protected at this point

That doesn't sound write, what do thay write in `target/debug`? AFAIK those test cases should be run in `cli/tests` as CWD

**Comment by 06kellyjac** [comment]
Thanks for merging!

---

> Benchmarks are only run on Ubuntu so it shouldn't be a problem
It's not clear in my comment but I'm generally referring to running all the tests in the nix package. I can skip it for aarch64 builds if necessary.
I'll probably end up always skipping it anyway as it's just for recording right? No kind of target & failure?

> Do they download from deno.land? If so we should change that to use local HTTP server (tools/http_server.py)
Yeah.
Local HTTP server sounds good, would it be worth waiting for the refactor?
https://github.com/denoland/deno/blob/42a04ea08a09e29417518524cbe2245c165e6f55/std/testing/bench_test.ts#L36-L43

> That doesn't sound write, what do thay write in `target/debug`? AFAIK those test cases should be run in `cli/tests` as CWD
Ah I might be wrong on that, I'll need to do some more checking. But pretty much the whole dir is read-only so writing to CWD isn't allowed :sweat_smile:
An env var override and/or pointing to `/tmp` would help with those ones

---

On my version bump to v1.0.5 I've disabled the tests because I've been getting cases where the output that's expected of another test appears.
Which is reeeally weird because it's using static read-only files for the code and dependencies, no previous cache in `target`, etc.)
With this PR adding some more logs I might be able to get to the bottom of it.
I'm thinking it might be some kind of concurrency thing with stdout but it's also odd I haven't seen these issues in the actions or in other peoples builds. Spooky

**Pull request merged**